Upgrade to C++17, fixing macOS build

Aquilum opened the pull request:
These changes were needed to build on macOS Catalina. For the largest part, it's just an upgrade to C++17 and the namespace changes that go with that, but also includes re-adding <sstream> to main.

There's an additional change to drop the filesystem library link instruction for APPLE platforms. This is a byproduct of the more recent compilers used in later macOS releases, but is probably better guarded on compiler versions than platform identifiers.

Aquilum committed:
cmake: Fixup macOS build

Aquilum committed:
main: C++17 update

Re-introduce sstream, as that's where stringstream is defined,
and replace the 'experimental' namespace, as C++17 introduced
filesystem in earnest.

Aquilum committed:
cmake: upgrade to c++17

jmccl commented on main/main.cpp:
```diff
@@ -3,8 +3,9 @@
 #include <cerrno>
 #include <fstream>
 #include <iostream>
+#include <sstream>
```
Interesting. Not sure why this compiles without this on linux.

Aquilum replied:
My best suggestion is that the different C++ standard libraries have different include trees. You might be able to reproduce it by linking against libcxx (via llvm/clang) rather than GCC (and I, vice versa, by trying libstdc++)

jmccl replied:
Yeah, I assume that's it, but weird.

jmccl commented on lib/CMakeLists.txt:
```diff
@@ -5,6 +5,6 @@ add_library(acme_lw ${source})
-target_include_directories(acme_lw PUBLIC ${CMAKE_CURRENT_SOURCE_DIR} ${CMAKE_CURRENT_SOURCE_DIR}/../internal )
+target_include_directories(acme_lw PUBLIC ${CMAKE_CURRENT_SOURCE_DIR} ${CMAKE_CURRENT_SOURCE_DIR}/../internal ${OPENSSL_INCLUDE_DIR})
 
 target_link_libraries(acme_lw PRIVATE crypto)
```
I tried building this on a Mac and it wouldn't link until I replaced 'crypto' with ${OPENSSL_LIBRARIES}. You didn't need to do that, or equivalent?

Aquilum replied:
Interesting. No, I didn't have to. My CMake install found OpenSSL via homebrew:

`OPENSSL_CRYPTO_LIBRARY           /usr/local/opt/openssl/lib/libcrypto.dylib
 OPENSSL_INCLUDE_DIR              /usr/local/opt/openssl/include
 OPENSSL_SSL_LIBRARY              /usr/local/opt/openssl/lib/libssl.dylib`

So I suspect it worked by accident. I'll make the change to use the FindOpenSSL macro as you suggest.

jmccl replied:
Thanks, that would be great. I used ports rather than brew and I assume that's the difference.

jmccl reviewed (commented):
Thanks for the PR. 

I'd been putting off moving to 17 just to support older compilers, but at some point it makes sense and I suppose that's now. Can you address my comment about linking at line 10 of lib/CMakeList.txt?

Aquilum committed:
cmake: Replace explicit crypto library link

As we're using the find_package(...) to locate OpenSSL,
let's use the library path it locates.

Built on macOS Catalina w/ CMake 3.17.3
